### ISSUE TITLE
Unexpected result in matchTemplate for TM_CCOEFF_NORMED

### ISSUE THREAD

**Issue by PkLab** [Itseez/opencv#5688]
When the template is flat (all pixels have same value) the result of TM_CCOEFF_NORMED is 0/0 but matchTemplate returns 1.
Looking at [the formula](http://docs.opencv.org/3.0.0/df/dfb/group__imgproc__object.html#ga3a7850640f1fe1f58fe91a2d7583695d), `T` is the template:
`T'(x,y) = T(x,y) - mean(T)`

If T(x,y)=C for each x,y than mean(T)=C and you will have 
`T'(x,y) = C - C = 0` for each x,y

the formula will be:
`R = 0 * I / sqrt(0 * I^2)` that is 0/0 !=1 

This is true even if the Image patch under running template is constant.

In this case matchTemplate should returns undefined even if 0 has more sense

here is [the issue](https://github.com/Itseez/opencv/blob/master/modules/imgproc/src/templmatch.cpp#L931)
here is simple test code:
```cpp
Mat img,templ;
Scalar C = Scalar::all(64); //some rand number
img = imread("some_image.png");
templ = C + Mat::zeros(10,10,img.type());  // flat template
int result_cols = img.cols - templ.cols + 1;
int result_rows = img.rows - templ.rows + 1;
resultMatching.create(result_rows, result_cols, CV_32FC1);
// TM_CCOEFF returns all 0 as expected
matchTemplate(img, templ, resultMatching, TM_CCOEFF);
// TM_CCOEFF returns all 1 despite of it has sam numerator as TM_CCOEFF
matchTemplate(img, templ, resultMatching, TM_CCOEFF_NORMED);
```

**Comment by snosov1** [comment]
I don't think providing "undefined" values over "all ones" has much merit. In either case the result makes all the positions "equal".

However, my thinking is that the evaluation could be slightly modified to provide somewhat helpful result. Current implementation of CCOEFF goes like "you can't match uniform templates", because if you try to - `T'` will always be equal to zero and the result will be all zeros, too.

But if we slightly modify the implementation by adding a small tolerance after subtracting the mean value, the result becomes somewhat meaningful - non-uniform areas will have lower CCOEFF values.

I've uploaded the naive implementation [here](https://gist.github.com/snosov1/6f5b971552e72140090f9b5908ab2bdb#file-gistfile1-txt-L47).  `matchTemplate` implementation can also be easily modified in a similar fashion. 

I can submit a PR if @vpisarev approves the proposal.

Also, while digging into this, I noticed that matchTemplate result is significantly different compared to the naive implementation. Consider the example image that is generated by the provided gist (with naive implementation the result is "all zeros"):

![mt](https://cloud.githubusercontent.com/assets/2784626/16807705/2639e5e4-4922-11e6-869f-291a1e91db73.png)

@vpisarev, maybe you could comment whether this is a known behavior.

**Comment by PkLab** [comment]
I think that result is mathematically undefined, also for the case of `TM_CCOEFF` thus methods are not applicable if image region or template or both are const (stdev==0)

Thinking about logic, because we are looking at similarity:

- If both template AND image region under the template are const 
  - if template=region=V than result should be 1
  - if template=T and region=R than some normalized similarity between R and T might be returned

 ```cpp
   diff = abs(T-R);
   //normalize
   result = 1 - diff / max(T,R);
 ```
- If region is const=R, than a value close to 1 might be returned if template is almost const with a mean value ~=R ... something like this

 ```cpp
T = mean(template);
diff = abs(mean(T - R) + stdev(template)/2.0;
//normalize
result = 1 - diff / max(T,R);
```
  or specular case when the template is const=T.

Summarizing, if template OR region is const, might be:

 ```cpp
T = mean(template);
R = mean(region);
diff = abs(T - V) + (stdev(region) + stdev(template) / 2.0;
//normalize
result = 1 - diff / max(T,R);
```
A computationally heavy but it could be an idea

**Comment by snosov1** [comment]
> I think that result is mathematically undefined

Yes, it's undefined mathematically. But we're not entirely mathematicians here (don't get me started on the floating point arithmetic =) ). So, my point is that it might be _practically_ viable to allow such cases.

The reasoning is - all the values we work with contain some amount of noise (with the simplest one being a quantization noise). So, it's good for any algorithm to be "stable" to such small changes in input values. Here we have a "singularity point" (a corner case) when we divide by zero and it kind of has a qualitative effect on the result. But if we artificially introduce a small amount of noise (e.g. add 1e-7 to mean and stdev values, which is much smaller than even a quantization noise), it suddenly makes the results meaningful again. 

This trick is a common practice in both CV and CG to handle "division-by-zero" cases (it's also a simple and efficient one).

**Comment by PkLab** [comment]
My point is result correctness or expected result in such cases. 

Take this image  and template:
![img](https://cloud.githubusercontent.com/assets/15944968/16833651/84798a9a-49b0-11e6-96e7-1833d0eab930.png) ![templ](https://cloud.githubusercontent.com/assets/15944968/16833653/875e8b0c-49b0-11e6-8695-cce0c953f77c.png)

below are results:
We know that ccoeff_normed returns all 1
Your implementation (normalized):
![snccoeff](https://cloud.githubusercontent.com/assets/15944968/16833659/8e006c6e-49b0-11e6-842d-a28f2cfb49d7.png)

My (revised) idea:
![pkccoeff](https://cloud.githubusercontent.com/assets/15944968/16833674/9f6d2dd4-49b0-11e6-98bd-3d8de79289a7.png)

**Comment by PkLab** [comment]
I've  just modified your ccoeff implementation so to get a ccoeff_normed. I've to say that result in this case is valuable... see below
![snccoeffnormed](https://cloud.githubusercontent.com/assets/15944968/16836484/fc2e78c2-49bd-11e6-98d0-1520fdc670ce.png)

non normed version has still issue also with my idea

**Comment by snosov1** [comment]
Yeah, that's what I've just wanted to post. 

So, you agree that adding small tolerance is a viable option?

**Comment by snosov1** [comment]
Ok, then. We'll wait for a blessing from @vpisarev on this, and also some comments on the difference of the matchTemplate result compared to naive implementation.

**Comment by PkLab** [comment]
OK even if the trick seems do not solve the issue for non normed version of CCOEFF

**Comment by vpisarev** [comment]
@snosov1, I bless any solution what would pass all the tests and would produce basically the same or very similar results in the case when the template is not "flat". If I understood your idea correctly, you suggest to slightly modify each component of the sum. Maybe, there is a better solution. At least, if you add some `eps`, defined as:

<pre>
double eps = flat_template ? 0. : 1e-5;
</pre>

where flat_template can be set by checking `templNorm`, that would give absolutely the same results in the case of non-flat template.

**Comment by snosov1** [comment]
Got it! Will implement something along those lines.

And what do you think about matchTemplate producing, let's say, "imperfect" results for artificial input?

In the [uploaded code sample](https://gist.github.com/snosov1/1a0a71a3261f9daa82a6b37dd1ee8348) matchTemplate should produce an all-zero output (and the naive implementation does produce that). However, it actually produces an image with a lot of non-zero (up to 0.140625) values. Here's the **mask** of the output:

![](https://cloud.githubusercontent.com/assets/2784626/16807705/2639e5e4-4922-11e6-869f-291a1e91db73.png)

Is that know and/or intended? Should I look into that deeper?

**Comment by PkLab** [comment]
Issue happens if template **OR** image region under template is flat. Thus tests should be performed with both cases.  Because `img_region` is a running ROI  a "flat test" would require to calculate additional N stdev or similar. N=result.Size();

BTW snosov1's implementation avoid the "flat test" and should works fine with CCOEFF_NORMED  if `template AND img_region >> epslion` but still produce nonsense result with CCOEFF (if region or template is flat).

I think that @vpisarev  is talking about regression tests.  @snosov1 why you don't prepare a branch on your fork including adequate test cases for https://github.com/opencv/opencv/blob/master/modules/imgproc/test/test_templmatch.cpp. This would help us to go on.

**Comment by snosov1** [comment]
After discussion with @vpisarev we agreed that the noisy output on artificial input is expected because of the way CCORR is implemented (using DFT). Because of this, adding a small epsilon doesn't make any significant difference in the output (it only makes a difference for a straightforward implementation). 

So, it kind of falls back to the original question whether the result for a flat template should be undefined. My opinion is the same - it doesn't have any practical difference, let's leave it as is.

**Comment by z6491679** [comment]
good issue，I also meet this problem. Although the probability of issue appearing is almost non-existent，the test engineer can pick you up.
